File: Android/app/src/main/java/com/jiny/catchtherule/data/MinigameRankingService.kt
package com.jiny.catchtherule.data

import android.content.Context
import com.jiny.catchtherule.core.model.GameMode
import com.jiny.catchtherule.core.model.RankEntry
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL

/**
 * Minigame 서버(https://duo.jiny.shop)의 CatchTheRule 랭킹 API 결합.
 *
 *   POST /api/catchtherule/scores       { nickname, score, mode, deviceId } -> { rank }
 *   GET  /api/catchtherule/leaderboard?mode=timeAttack&limit=100 -> { entries: [...] }
 *
 * 로그인이 없으므로 인증 없이 호출하고, 기기당 1행 유지를 위해
 * 로컬에 영속화한 deviceId(UUID)를 함께 보낸다.
 */
class MinigameRankingService(context: Context) : RankingService {

    private val baseUrl = "https://duo.jiny.shop"
    private val deviceId: String = CtrDevice.id(context)   // 문의 서비스와 동일한 값 공유
    private val country: String? = java.util.Locale.getDefault().country.takeIf { it.length == 2 }
    private val json = Json { ignoreUnknownKeys = true }

    override suspend fun submit(score: Int, nickname: String, mode: GameMode): Int =
        withContext(Dispatchers.IO) {
            val body = json.encodeToString(SubmitBody(nickname, score, mode.key, deviceId, country))
            val text = request("POST", "/api/catchtherule/scores", body)
            json.decodeFromString<SubmitResponse>(text).rank
        }

    override suspend fun leaderboard(mode: GameMode): List<RankEntry> =
        withContext(Dispatchers.IO) {
            val text = request("GET", "/api/catchtherule/leaderboard?mode=${mode.key}&limit=100", null)
            json.decodeFromString<LeaderboardResponse>(text).entries.map {
                RankEntry(rank = it.rank, nickname = it.nickname, score = it.score, country = it.country)
            }
        }

    // MARK: - HTTP

    private fun request(method: String, path: String, body: String?): String {
        val conn = (URL(baseUrl + path).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 10_000
            readTimeout = 10_000
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
            }
        }
        try {
            if (body != null) conn.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            val code = conn.responseCode
            val stream = if (code in 200..299) conn.inputStream else conn.errorStream
            val text = stream?.bufferedReader(Charsets.UTF_8)?.use { it.readText() } ?: ""
            if (code !in 200..299) throw RuntimeException("HTTP $code: $text")
            return text
        } finally {
            conn.disconnect()
        }
    }

    @Serializable
    private data class SubmitBody(
        val nickname: String,
        val score: Int,
        val mode: String,
        val deviceId: String,
        val country: String?,
    )

    @Serializable
    private data class SubmitResponse(val rank: Int)

    @Serializable
    private data class LeaderboardResponse(val entries: List<Entry>) {
        @Serializable
        data class Entry(val rank: Int, val nickname: String, val score: Int, val country: String? = null)
    }
}

File: iOS/CatchTheRule/Services/MinigameRankingService.swift
import Foundation

/// Minigame 서버(https://duo.jiny.shop)의 CatchTheRule 랭킹 API 결합.
///
///   POST /api/catchtherule/scores       { nickname, score, mode, deviceId } -> { rank }
///   GET  /api/catchtherule/leaderboard?mode=timeAttack&limit=100 -> { entries: [...] }
///
/// 로그인이 없으므로 인증 헤더 없이 호출하고, 기기당 1행 유지를 위해
/// 로컬에 영속화한 deviceId(UUID)를 함께 보낸다.
final class MinigameRankingService: RankingService {

    private let baseURL = URL(string: "https://duo.jiny.shop")!
    private let session: URLSession
    private let deviceID: String
    private let country: String?

    init(session: URLSession = .shared) {
        self.session = session
        self.deviceID = Self.resolveDeviceID()
        self.country = Self.resolveCountry()
    }

    // MARK: - RankingService

    func submit(score: Int, nickname: String, mode: GameMode) async throws -> Int {
        let url = baseURL.appendingPathComponent("api/catchtherule/scores")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(
            SubmitBody(nickname: nickname, score: score, mode: mode.serverKey,
                       deviceId: deviceID, country: country)
        )

        let (data, response) = try await session.data(for: request)
        try Self.check(response, data)
        return try JSONDecoder().decode(SubmitResponse.self, from: data).rank
    }

    func leaderboard(mode: GameMode) async throws -> [RankEntry] {
        var comps = URLComponents(
            url: baseURL.appendingPathComponent("api/catchtherule/leaderboard"),
            resolvingAgainstBaseURL: false
        )!
        comps.queryItems = [
            URLQueryItem(name: "mode", value: mode.serverKey),
            URLQueryItem(name: "limit", value: "100"),
        ]
        var request = URLRequest(url: comps.url!)
        request.httpMethod = "GET"

        let (data, response) = try await session.data(for: request)
        try Self.check(response, data)
        let decoded = try JSONDecoder().decode(LeaderboardResponse.self, from: data)
        return decoded.entries.map {
            RankEntry(rank: $0.rank, nickname: $0.nickname, score: $0.score, country: $0.country)
        }
    }

    // MARK: - Helpers

    private static func check(_ response: URLResponse, _ data: Data) throws {
        guard let http = response as? HTTPURLResponse else { return }
        guard (200..<300).contains(http.statusCode) else {
            let message = (try? JSONDecoder().decode(ErrorResponse.self, from: data))?.error
                ?? "HTTP \(http.statusCode)"
            throw RankingError.server(message)
        }
    }

    /// 기기 식별자(UUID). 문의 서비스와 동일한 값을 공유한다.
    private static func resolveDeviceID() -> String { CTRDevice.id }

    /// 기기 지역(ISO 3166-1 alpha-2, 예: "KR"). 랭킹 국기 표시용.
    private static func resolveCountry() -> String? {
        if #available(iOS 16, *) {
            return Locale.current.region?.identifier
        }
        return Locale.current.regionCode
    }

    // MARK: - DTO

    private struct SubmitBody: Encodable {
        let nickname: String
        let score: Int
        let mode: String
        let deviceId: String
        let country: String?
    }
    private struct SubmitResponse: Decodable {
        let rank: Int
    }
    private struct LeaderboardResponse: Decodable {
        let entries: [Entry]
        struct Entry: Decodable {
            let rank: Int
            let nickname: String
            let score: Int
            let country: String?
        }
    }
    private struct ErrorResponse: Decodable {
        let error: String
    }
}
